File: tests/xcalablemp/global-view/loop/F/reduction_max2.f90
program main
  include 'xmp_lib.h'
  integer,parameter:: N=10
  integer random_array(N*N), ans_val, result
  integer a(N,N), sa
  real*8  b(N,N), sb
  real*4  c(N,N), sc
  real tmp(N,N)
!$xmp nodes p(4,*)
!$xmp template t(N,N)
!$xmp distribute t(block,block) onto p
  
  result = 0
  call random_number(tmp)
  do j=1, N
     do i=1, N
        l = (j-1)*N+i
        random_array(l) = int(tmp(i,j) * 10000)
     end do
  end do

!$xmp loop (i,j) on t(i,j)
  do j=1, N
     do i=1, N
        l = (j-1)*N+i
        a(i,j) = random_array(l)
        b(i,j) = dble(random_array(l))
        c(i,j) = real(random_array(l))
     enddo
  enddo
         
  ans_val = 0
  do j=1, N
     do i=1, N
        l = (j-1)*N+i
        ans_val = max(ans_val, random_array(l))
     enddo
  enddo
  
  sa = 0
  sb = 0.0
  sc = 0.0

!$xmp loop (i,j) on t(i,j) reduction(max: sa, sb, sc)
  do j=1, N
     do i=1, N
        sa = max(sa, a(i,j))
        sb = max(sb, b(i,j))
        sc = max(sc, c(i,j))
     enddo
  enddo

  if( sa .ne. ans_val .or. sb .ne. dble(ans_val) .or. sc .ne. real(ans_val) ) then
     result = -1
  endif

!$xmp reduction(+:result)
!$xmp task on p(1,1)
  if( result .eq. 0 ) then
     write(*,*) "PASS"
  else
     write(*,*) "ERROR"
     call exit(1)
  endif
!$xmp end task
      
end program main

File: tests/xcalablemp/global-view/loop/C-Round/reduction_max2.c
#include<xmp.h>
#include<stdio.h>  
#include<stdlib.h> 
static const int N=1000;
int random_array[1000000],ans_val=0;
int a[1000][1000],sa=0;
double b[1000][1000],sb=0.0;
float c[1000][1000],sc=0.0;
#pragma xmp nodes p(4,*)
#pragma xmp template t(0:N-1,0:N-1)
#pragma xmp distribute t(block,block) onto p
int i,j,k,l,result=0;

int main(void)
{
  srand(0);
  for(i=0;i<N*N;i++)
    random_array[i] = rand();
    
#pragma xmp loop (j,i) on t(j,i)
  for(i=0;i<N;i++){
    for(j=0;j<N;j++){
      l = j*N+i;
      a[i][j] = random_array[l];
      b[i][j] = (double)random_array[l];
      c[i][j] = (float)random_array[l];
    }
  }

  for(i=0;i<N;i++){
    for(j=0;j<N;j++){
      l = j*N+i;
      if(random_array[l]>ans_val){
	ans_val=random_array[l];
      } 
    }
  }
      
#pragma xmp loop (j,i) on t(j,i) reduction(max:sa)
  for(i=0;i<N;i++)
    for(j=0;j<N;j++)
      if(a[i][j]>sa)
	sa=a[i][j];

#pragma xmp loop (j,i) on t(j,i) reduction(max:sb)
  for(i=0;i<N;i++)
    for(j=0;j<N;j++)
      if(b[i][j]>sb)
	sb=b[i][j];

#pragma xmp loop (j,i) on t(j,i) reduction(max:sc)
  for(i=0;i<N;i++)
    for(j=0;j<N;j++)
      if(c[i][j]>sc)
	sc=c[i][j];

  if( (sa != ans_val) || (sb != (double)ans_val) || (sc != (float)ans_val) )
    result = -1;

#pragma xmp reduction(+:result)
#pragma xmp task on p(1,1)
  {
    if(result == 0){
      printf("PASS\n");
    }
    else{
      fprintf(stderr, "ERROR\n");
      exit(1);
    }
  }
  return 0;
}
